src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -3925,6 +3940,121 @@ void EmitStackPop()
             }
         }
 
+        protected void EmitScan(DynamicMethod findFirstCharMethod, DynamicMethod goMethod)
+        {
+            LocalBuilder bump = DeclareInt32();
+            LocalBuilder stoppos = DeclareInt32();
+            Label returnLabel = DefineLabel();
+
+            Label notRightToLeft = DefineLabel();
+            // int bump = 1
+            Ldc(1);
+            Stloc(bump);
+
+            // int stoppos = text.Length
+            _ilg!.Emit(OpCodes.Ldarga_S, 1);
+            Call(s_spanGetLengthMethod);
+            Stloc(stoppos);
+
+            // if (runregex.RightToLeft)
+            // {
+            Ldthisfld(s_runregexField);
+            Callvirt(s_regexGetRightToLeft);
+            BrfalseFar(notRightToLeft);
+
+            // bump = -1;
+            // stoppos = 0;
+            // }
+            Ldc(-1);
+            Stloc(bump);
+            Ldc(0);
+            Stloc(stoppos);
+            MarkLabel(notRightToLeft);
+
+            // while (true)
+            Label whileLoopEnd = DefineLabel();
+            Label whileLoopBody = DefineLabel();
+            MarkLabel(whileLoopBody);
+
+            // if (FindFirstChar(text))
+            Label afterFindFirstCharLabel = DefineLabel();
+            Ldthis();
+            Ldarg_1();
+            Call(findFirstCharMethod);
+            BrfalseFar(afterFindFirstCharLabel);
+
+            // CheckTimeout();
+            Ldthis();
+            Call(s_checkTimeoutMethod);
+
+            // Go(text);
+            Ldthis();
+            Ldarg_1();
+            Call(goMethod);
+
+            // if (runmatch!._matchcount[0] > 0)
+            //   return;
+            Label afterSuccessMatchLabel = DefineLabel();
+            Ldthisfld(s_runmatchField);
+            _ilg!.Emit(OpCodes.Ldfld, s_matchMatchCountField);
+            Ldc(0);
+            LdelemI4();
+            Ldc(0);
+            _ilg!.Emit(OpCodes.Cgt);
+            BrfalseFar(afterSuccessMatchLabel);
+            BrFar(returnLabel);
+
+            // runtrackpos = runtrack!.Length;
+            MarkLabel(afterSuccessMatchLabel);
+            Ldthis();
+            Ldthisfld(s_runtrackField);
+            Ldlen();
+            _ilg!.Emit(OpCodes.Conv_I4);
+            Stfld(s_runtrackposField);
+
+            // runtrackpos = runstack!.Length;
+            Ldthis();
+            Ldthisfld(s_runstackField);
+            Ldlen();
+            _ilg!.Emit(OpCodes.Conv_I4);
+            Stfld(s_runstackposField);
+
+            // runcrawlpos = runcrawl!.Length;
+            Ldthis();
+            Ldthisfld(s_runcrawlField);
+            Ldlen();
+            _ilg!.Emit(OpCodes.Conv_I4);
+            Stfld(s_runcrawlposField);
```

Comment:
Same questions as on the source gen about whether we actually need to do this resetting.

(We should strive to keep the RegexCompiler implementation and the source generator implementation 1:1 as much as possible, only diverging in places where we're forced to do something special because of C# vs IL (e.g. complications from scoping in C#) or where emitting C# enables further optimizations (e.g. emitting a switch that the C# compiler can further optimize with its heuristics).